---
output:
  html_document: default
  pdf_document: default
---

# GWAS in PGx

Greg McInnes
May 23, 2021


This notebook contains the code to reproduce figures for the manuscript "Genome-wide Association Studies in Pharmacogenomics". It also contains code and figures for a more exploratory analysis of the data in [GWAS Catalog](https://www.ebi.ac.uk/gwas/) with an emphasis on pharmacogenomics.


## Setup

# Set your working directory. **Change this to work on your own machine**
```{r setup}
working_directory = "/Users/gregmcinnes/projects/gwas_in_pgx/repo/"
knitr::opts_knit$set(root.dir = working_directory)
```


### Load packages
```{r, echo=FALSE}
library(tidyverse)
library(lubridate)
library(ggrepel)
library(scales)
```


### Read in GWAS Catalog data

The files from GWAS Catalog were retrieved using the jupyter notebook from the publication ["A sciometric review of genome-wide assication studies"](https://github.com/crahal/GWASReview).  These files can also be obtained directly from the [GWAS Catalog](https://www.ebi.ac.uk/gwas/docs/file-downloads).

```{r}
# Load the files
cat_map <- read.table("data/Cat_Map.tsv", sep="\t", header=T, quote="")
cat_full <- read.table("data/Cat_Full.tsv", sep="\t", header=T, quote="")
cat_anc <- read.table("data/Cat_Anc.tsv", sep="\t", header=T, quote="")
cat_study <- read.table("data/Cat_Stud.tsv", sep="\t", header=T, quote="")

# Reformat the date columns
cat_full <- cat_full %>%
  mutate(DATE = as.Date(DATE),
         DATE.ADDED.TO.CATALOG = as.Date(DATE.ADDED.TO.CATALOG))

cat_study <- cat_study %>%
  mutate(DATE = as.Date(DATE),
         DATE.ADDED.TO.CATALOG = as.Date(DATE.ADDED.TO.CATALOG))
```

Load the annotated study details from the curated PGx GWAS
```{r}
study_details <- read.table("data/pgx_gwas_description_current.tsv", header=T, sep="\t", fill=T, quote='', comment.char = "")

# Add the publication date to the curated data
study_details <- study_details %>%
  left_join(cat_full %>% select(PUBMEDID, DATE) %>% unique %>% rename(pubmed_id = PUBMEDID)) %>%
    mutate(year = year(DATE)) 

study_details %>% head
```

Load ATC code descriptions
```{r}
atc_codes <- read.table("data/atc_map.tsv", sep="\t", header=T) %>% rename(atc_description=Preferred.Label)
atc_codes %>% head
```

Set the plot theme. Derived from [here](https://rpubs.com/Koundy/71792).

```{r}
theme_Publication <- function(base_size=14, base_family="Helvetica") {
      library(grid)
      library(ggthemes)
      (theme_foundation(base_size=base_size, base_family=base_family)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(face = "bold",size = rel(1)),
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               #legend.position = "bottom",
               #legend.direction = "horizontal",
               #legend.key.size= unit(0.2, "cm"),
               #legend.margin = unit(0, "cm"),
               #legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
          ))
      
}

scale_fill_Publication <- function(...){
      library(scales)
      discrete_scale("fill","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)

}

scale_colour_Publication <- function(...){
      library(scales)
      discrete_scale("colour","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)

}
```


## PGx Analysis

This first part has some general exploratory analysis and figures to better understand the data

### General publication trends

Number of published PGx GWAS papers over time

```{r}
study_details %>%
  unique %>%
  ggplot(aes(year)) +
  geom_bar() +
  #theme_minimal() +
  theme_Publication(base_size = 12) +
  ggtitle("PGx GWAS publications over time") +
  ylab("Number of publications") +
  xlab("Year") +
  theme(panel.grid.major.x  = element_blank())

ggsave("figures/pgx_gwas_count.png", width=6, height=4)
```

Show the counts of papers over time, sorted.
```{r}
study_details %>%
  unique %>%
  mutate(ATC_chapter = substr(atc, 1, 1)) %>%
  filter(ATC_chapter != "") %>%
  filter(!is.na(year)) %>%
  group_by(year) %>%
  tally %>%
  arrange(-n)
```

Plot all GWAS over time. Color PGx and all other GWAS separately.

```{r}
cat_full %>%
  select(PUBMEDID, DATE, JOURNAL) %>%
  unique %>%
  mutate(DATE = as.Date(DATE)) %>%
  mutate(year = year(DATE)) %>%
  filter(year < 2021) %>%
  mutate(type = case_when(PUBMEDID %in% study_details$pubmed_id ~ 'PGx',
                   T ~ 'Other GWAS')) %>%
  ggplot(aes(as.factor(year), fill=type)) +
  geom_bar(stat="count", alpha=0.9) +
  ggtitle("GWAS Catalog Entries") +
  ylab("Number of publications") +
  xlab("Year")  +
  theme_Publication(base_size = 12) +
  scale_fill_Publication() +
  theme(panel.grid.major.x  = element_blank()) +
  labs(fill="")
  
ggsave("figures/gwas_catalog_counts.png", width=6, height=4)
```


Calculate the proportion of studies each year that were PGx each year

```{r}
cat_study %>%
  select(PUBMEDID, DATE, JOURNAL) %>%
  unique %>%
  mutate(DATE = as.Date(DATE)) %>%
  mutate(year = year(DATE)) %>%
  filter(year < 2021) %>%
  mutate(type = case_when(PUBMEDID %in% study_details$pubmed_id ~ 'PGx',
                   T ~ 'Other')) %>%
  group_by(year, type) %>%
  tally %>%
  pivot_wider(names_from = type, values_from = n) %>%
  mutate(PGx = case_when(is.na(PGx) ~ 0.0, T~as.numeric(PGx))) %>%
  mutate(fraction = (PGx / Other)) %>%
  ggplot(aes(year, fraction)) +
    geom_point(size=2) +
  geom_line() +
  theme_Publication(base_size = 12) +
  scale_colour_Publication() +
  labs(fill="") +
  ggtitle("Percent of PGx-related GWAS Catalog entries") +
  ylab("Percent") +
  xlab("Year") +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1))
  

ggsave("figures/pgx_gwas_fraction.png", width=6, height=4)
```



Determine what percent PGx GWAS are of the total

```{r}
total_gwas_count <- cat_study %>%
  select(PUBMEDID) %>%
  unique %>%
  nrow

pgx_gwas_count = cat_study %>%
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  select(PUBMEDID) %>%
  unique %>%
  nrow

(pgx_gwas_count / total_gwas_count) * 100
```

~8.2% of all GWAS papers in GWAS Catalog are PGx related




### Drug classes

Which drug classes are the most studied? Add ATC codes to each study and determine the frequency of each ATC code.

```{r message=FALSE, warning=FALSE}
study_details %>%
  mutate(ATC = substr(atc, 1, 3)) %>%
  left_join(atc_codes %>% select(ATC, atc_description)) %>%
  ungroup %>%
  filter(ATC != "") %>%
  group_by(ATC) %>%
  tally %>%
  ggplot(aes(reorder(ATC, n), n)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  theme_bw() +
  ggtitle("Number of GWAS for ATC Level 3") +
  theme_Publication() +
  scale_colour_Publication() +
  theme(panel.grid.major.y  = element_blank()) +
  xlab("ATC code") +
  ylab("Count")

ggsave("figures/atc_gwas_counts.png",
       width = 5,
       height = 6.5)
```

Color the ATC codes each year
```{r message=FALSE, warning=FALSE}
study_details %>%
  mutate(ATC = substr(atc, 1, 1)) %>%
  left_join(atc_codes %>% select(ATC, atc_description)) %>%
  ggplot(aes(year, fill=atc_description)) +
  geom_bar() +
  theme_bw() +
  ylab("Count") +
  xlab("Year") +
  ggtitle("PGx GWAS for each ATC Chapter")
```


### Significant associatoins in PGx GWAS

What fraction of papers have significant hits
```{r}
(cat_full %>%
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  filter(P.VALUE < 5e-08) %>%
  select(PUBMEDID) %>%
  unique %>%
  nrow) / nrow(study_details %>% select(pubmed_id) %>% unique)
```

Only 44% of papers have a significant hit


#### Has the fraction of papers with significant hits changed over time

```{r message=FALSE, warning=FALSE}
papers_with_significant_hits <- cat_full %>%
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  filter(P.VALUE < 5e-08) %>%
  select(PUBMEDID) %>%
  unique %>%
  pull

study_details %>%
  mutate(
    significant = case_when(
      pubmed_id %in% papers_with_significant_hits ~ "Significant hits",
      T ~ "No significant hits"
    )
  ) %>%
  unique %>%
  ggplot(aes(year, fill = significant)) +
  geom_bar(position = "fill") +
  theme_bw() +
  ggtitle("Fraction of PGx GWAS Publications with Significant Hits") +
  ylab("Count") +
  xlab("Year") +
  theme_Publication(base_size = 12) +
  scale_fill_Publication() +
  labs(fill = "")

ggsave("figures/pgx_gwas_fraction_significant.png",
       width = 6,
       height = 4)
```

How many drug classes / drugs have significant hits
```{r message=FALSE, warning=FALSE}
study_details %>%
  mutate(
    significant = case_when(
      pubmed_id %in% papers_with_significant_hits ~ "Significant hits",
      T ~ "No significant hits"
    )
  ) %>%
  mutate(ATC = substr(atc, 1, 3)) %>%
  unique %>%
  left_join(atc_codes %>% select(ATC, atc_description)) %>%
  ggplot(aes(ATC, fill = significant)) +
  geom_bar() +
  coord_flip() +
  theme_bw() +
  ggtitle("PGx GWAS by ATC Code") +
  ylab("Count") +
  xlab("ATC") +
  theme_Publication(base_size = 12) +
  scale_fill_Publication() +
  labs(fill = "")
```

That looks good, but lets sort it so it's a bit more interpretable.
```{r message=FALSE, warning=FALSE}
atc_totals <- study_details %>%
  mutate(
    significant = case_when(
      pubmed_id %in% papers_with_significant_hits ~ "Significant hits",
      T ~ "No significant hits"
    )
  ) %>%
  mutate(ATC = substr(atc, 1, 3)) %>%
  left_join(atc_codes %>% select(ATC, atc_description)) %>%
  ungroup %>%
  filter(ATC != "") %>%
  group_by(ATC, significant) %>%
  tally %>%
  ungroup %>%
  group_by(ATC) %>%
  summarize(total = sum(n))

study_details %>%
  mutate(
    significant = case_when(
      pubmed_id %in% papers_with_significant_hits ~ "Significant hits",
      T ~ "No significant hits"
    )
  ) %>%
  mutate(ATC = substr(atc, 1, 3)) %>%
  left_join(atc_codes %>% select(ATC, atc_description)) %>%
  ungroup %>%
  filter(ATC != "") %>%
  group_by(ATC, significant) %>%
  tally %>%
  left_join(atc_totals) %>%
  ggplot(aes(reorder(ATC, total), n, fill = significant)) +
  geom_bar(stat = 'identity') +
  coord_flip() +
  theme_bw() +
  ggtitle("PGx GWAS by ATC Code") +
  ylab("Count") +
  xlab("ATC") +
  theme_Publication(base_size = 12) +
  scale_fill_Publication() +
  labs(fill = "")

ggsave("figures/atc_gwas_counts_w_hits.png",
       width = 5,
       height = 6.5)
```


Which types of endpoints are studied for each drug class?
```{r message=FALSE, warning=FALSE}
study_details %>%
  mutate(ATC = substr(atc, 1, 3)) %>%
  unique %>%
  left_join(atc_codes %>% select(ATC, atc_description)) %>%
  ggplot(aes(ATC, fill = phenotype)) +
  geom_bar() +
  coord_flip() +
  theme_bw() +
  ggtitle("PGx GWAS by ATC Code") +
  ylab("Count") +
  xlab("ATC")
```

Get all the significant hits from all PGx studies
```{r}
pgx_gwas_hits <- cat_full %>%
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  left_join(study_details %>% select(pubmed_id, atc, phenotype) %>% rename(PUBMEDID=pubmed_id)) %>%
  select(PUBMEDID, DATE, JOURNAL, CHR_ID, CHR_POS, SNPS, REPORTED.GENE.S., CONTEXT, P.VALUE, atc, phenotype) %>%
  rename(CHR=CHR_ID, BP=CHR_POS, P=P.VALUE, genes=REPORTED.GENE.S., rsid=SNPS) %>%
  mutate(ATC = substr(atc, 1, 3)) %>%
  left_join(atc_codes %>% select(ATC, atc_description)) %>%
  group_by(ATC, atc_description) %>%
    mutate(CHR_NUM=case_when(CHR == "X" ~ "23",
                           CHR == "12;12;12" ~ "12",
                           CHR == "2;2;2;2;2;2;2;2;2;2;2;2" ~ "2",
                           CHR == "5;5;5;5;5;5;5;5;5" ~ "5",
                           CHR == "8;8;8" ~ "8",
                           T ~ CHR)
         ) %>%
  mutate(CHR_NUM=as.numeric(CHR_NUM)) %>%
  select(-CHR) %>%
  rename(CHR=CHR_NUM) %>%
  #mutate(CHR)
  filter(!is.na(CHR)) %>%
  mutate(BP = as.numeric(BP)) %>%
  filter(!is.na(BP))
```


Count the number of studies and significant hits for each ATC chapter.
```{r message=FALSE, warning=FALSE}
hit_count <- pgx_gwas_hits %>%
  mutate(significant = P < 5e-08) %>%
  mutate(counter = 1) %>%
  group_by(ATC, atc_description, phenotype, CHR, BP) %>%
  unique %>%
  ungroup %>%
  group_by(ATC, atc_description, phenotype) %>%
  summarize(n_sig_hits = sum(significant))

study_count <- pgx_gwas_hits %>%
  mutate(significant = case_when(P < 5e-08 ~ 1, T ~ 0)) %>%
  mutate(counter = 1) %>%
  select(PUBMEDID, ATC, atc_description, phenotype, counter, significant) %>%
  group_by(PUBMEDID, ATC, atc_description, phenotype) %>%
  filter(significant == max(significant)) %>%
  unique %>%
  ungroup %>%
  group_by(ATC, atc_description, phenotype) %>%
  select(ATC, atc_description, phenotype, counter, significant) %>%
  summarize(n_studies = sum(counter),
            n_studies_sig = sum(significant))

```
Let's turn that into a figure so it's a little easier to look at. We'll focus on side effects and efficacy.

```{r message=FALSE, warning=FALSE}
atc_order <- study_count %>%
  ungroup %>%
  filter(phenotype %in% c("side effects", "efficacy")) %>%
  select(ATC, n_studies) %>%
  group_by(ATC) %>%
  summarize(n = sum(n_studies)) %>%
  arrange(n) %>%
  select(ATC) %>%
  tail(n = 25) %>%
  pull

study_count %>%
  filter(ATC %in% atc_order) %>%
  mutate(ATC = ordered(ATC, levels = atc_order)) %>%
  arrange(-n_studies) %>%
  filter(phenotype %in% c("side effects", "efficacy")) %>%
  mutate(n_insig = n_studies - n_studies_sig) %>%
  select(-n_studies) %>%
  pivot_longer(cols = c("n_studies_sig", "n_insig")) %>%
  mutate(plot_name = case_when(
    name == "n_studies_sig" ~ "Significant hits",
    T ~ "No significant hits"
  )) %>%
  ggplot(aes(ATC, value, fill = plot_name)) +
  geom_bar(stat = "identity") +
  facet_grid(. ~ phenotype) +
  coord_flip() +
  ylab("ATC Code") +
  ylab("Number of studies") +
  ggtitle("Number of PGx GWAS publications") +
  scale_fill_discrete(name = "Study findings") +
  theme_Publication(base_size = 12) +
  scale_fill_Publication() +
  labs(fill = "")

ggsave(
  "figures/atc_gwas_counts_w_hits_phenotype.png",
  width = 5,
  height = 6.5
)
```




### Novel Associations

In order to assess the number of novel associations discovered each year we will need to perform linkage disequilibrium (LD) pruning to ensure that we aren't double counting SNPs that are just tightly linked in LD. To do this we'll use the package LDlinkR which calculates LD based on 1000 Genomes.

To run this code you will need your own token to access the LDlinkR servers. You can find instructions to obtain that here:
https://cran.r-project.org/web/packages/LDlinkR/vignettes/LDlinkR.html

Alternatively you can load the pre-pruned data used in the manuscript.

*NOTE* This step takes a considerable amount of time to run. I only recommend doing this if you want to update pruning data for a fresh analysis in the future. You can load the pre-pruned SNPs using the following block and skip the pruning step if you want to replicate the data in the manuscript.

```{r}
# LOAD THIS FILE AND SKIP THE NEXT CODE BLOCK IF YOU DO NOT NEED TO RECOMPUTE PRUNING
all_pruned_data <- read.table("data/cummulative_associations.txt", header=T, sep="\t")
```


Load LDlinkR and set the token
```{r}
library(LDlinkR)
my_token = '446ccd8dc47a'
```

Define a function to perform LD pruning. We will keep only the SNP with the most significant association for each drug class.
```{r}
prune_snps <- function(snp_df) {
  if (nrow(snp_df) < 2) {
    return(c())
  }
  
  snp_list <-
    snp_df %>% ungroup %>% select(rsid) %>% unique %>% pull
  
  all_pops <- list_pop()$pop_code
  correlation_df <-
    LDmatrix(
      snp_list,
      pop =  all_pops,
      r2d = "r2",
      token = my_token,
      file = FALSE
    )
  
  correlation_matrix <-
    as.matrix(correlation_df %>% remove_rownames %>% column_to_rownames(var = "RS_number"))
  diag(correlation_matrix) <- 0
  
  rsids <- colnames(correlation_matrix)
  
  if (length(rsids) < 2) {
    return(c())
  }
  
  if (length(rsids) > 0) {
    ld_data <- data.frame()
    for (i in 1:length(rsids)) {
      rsid_1 = rsids[i]
      r1_p <-
        snp_df %>% filter(rsid == rsid_1) %>% group_by(rsid) %>% filter(P == min(P)) %>% select(P) %>% pull
      for (j in 1:length(rsids)) {
        rsid_2 = rsids[j]
        if (rsid_1 != rsid_2) {
          r2_p <-
            snp_df %>% filter(rsid == rsid_2) %>% group_by(rsid) %>% filter(P == min(P)) %>% select(P) %>% pull
          # add r1 and r2 to data frame, with the r2 value and the p.value and which one is the minimum
          r2 = correlation_matrix[rsid_1, rsid_2]
          new <-
            data.frame(
              rsid_1 = rsid_1,
              rsid_2 = rsid_2,
              r2 = r2,
              rsid_1_p = r1_p,
              rsid_2_p = r2_p
            )
          ld_data <- bind_rows(ld_data, new)
          
          
        }
      }
    }
    
    
    
    ld_data <- ld_data %>%
      filter(r2 > 0.5) %>%
      mutate(min_rsid = case_when(rsid_1_p <= rsid_2_p ~ rsid_1,
                                  rsid_2_p < rsid_1_p ~ rsid_2)) %>%
      mutate(discard = case_when(rsid_1_p <= rsid_2_p ~ rsid_2,
                                 rsid_2_p < rsid_1_p ~ rsid_1))
    
    discards <- ld_data %>%
      select(discard) %>%
      unique %>%
      pull
    
    
    ld_data %>%
      filter(!min_rsid %in% discards)
    
    print(discards)
    return(discards)
    
  } else {
    #print("No snps in LD")
    return(c())
  }
  
}

```




Ok, now what we will do is for each ATC group get the unique SNPs that have been discovered each year. Then each year also check for SNPs that have been discovered in previous years. Check for those after pruning.
```{r, message=FALSE, warning=FALSE, eval=FALSE}
pruning_data <- pgx_gwas_hits %>%
  filter(P <= 5e-08) %>%
  mutate(year = year(DATE)) %>%
  select(CHR, BP, rsid, year, ATC, P) %>%
  group_by(CHR, BP, ATC, year) %>%
  filter(P == min(P)) %>%
  filter(ATC != "") %>%
  unique

atc_groups <- pruning_data %>%
  ungroup %>%
  select(ATC) %>%
  arrange(ATC) %>%
  unique %>%
  pull

all_pruned_data <- data.frame()
for (i in 1:length(atc_groups)) {
  code = atc_groups[i]
  print(code)
  code_data <-
    pruning_data %>% filter(ATC == code) %>% arrange(CHR, BP)
  pruned_code_data <- data.frame()
  years <-
    code_data %>% ungroup %>% select(year) %>% arrange(year) %>% unique %>% pull
  for (j in 1:length(years)) {
    y = years[j]
    print(y)
    chromosomes <-
      code_data %>% ungroup %>% filter(year == y) %>% select(CHR) %>% unique %>% pull
    for (k in 1:length(chromosomes)) {
      chr = chromosomes[k]
      prune_data <- code_data %>% ungroup %>% filter(year == y, CHR == chr) %>% unique
      discards <- prune_snps(prune_data)
      # Remove LD SNPs
      prune_data <- prune_data %>% filter(!rsid %in% discards)
      # Remove previously identified SNPs
      prune_data <-
        prune_data %>% filter(!rsid %in% pruned_code_data$rsid)
      # Add to existings data
      pruned_code_data <- bind_rows(pruned_code_data, prune_data)
    }
  }
  all_pruned_data <- bind_rows(all_pruned_data, pruned_code_data)
}

```


Save the pruned data
```{r}
write.table(all_pruned_data, file="data/cummulative_associations.txt", row.names = F, quote=F, sep="\t")
```


Plot the number of newly discovered associations each year.
```{r}
all_pruned_data %>%
  group_by(year) %>%
  tally %>%
  mutate(cummulative_snps = cumsum(n)) %>%
  ggplot(aes(year, cummulative_snps)) +
  geom_point(size = 2) +
  geom_line() +
  scale_x_continuous(
    labels = c(2008, 2010, 2012, 2014, 2016, 2018, 2020),
    breaks = c(2008, 2010, 2012, 2014, 2016, 2018, 2020)
  ) +
  ylab("Cummulative SNP associations") +
  xlab("Year") +
  ggtitle("Cummulative PGx associations discovered through GWAS") +
  theme_bw() +
  theme_Publication(base_size = 12) +
  scale_fill_Publication() +
  labs(fill = "")
```


Plot the number of newly discovered associations each year for each drug class.
```{r}
top_atc_codes <- all_pruned_data %>%
  group_by(ATC) %>%
  tally %>%
  arrange(-n) %>%
  head(n = 10) %>%
  select(ATC) %>%
  pull

top_atc_codes <- c(top_atc_codes, "Other")

all_pruned_data %>%
  mutate(ATC = case_when(ATC %in% top_atc_codes ~ ATC, T ~ "Other")) %>%
  mutate(ATC = ordered(ATC, levels = top_atc_codes)) %>%
  ggplot(aes(as.factor(year), fill = ATC)) +
  geom_bar() +
  theme_minimal() +
  ylab("Year") +
  ylab("Count") +
  theme_Publication(base_size = 12) +
  scale_fill_brewer(palette = "Spectral") +
  labs(fill = "ATC") +
  ylab("Number of new association") +
  xlab("Year") +
  ggtitle("Number of newly discovered PGx associations each year")

ggsave(
  "figures/pgx_association_discovery_count.png",
  width = 6,
  height = 4
)
```

Total number of significant associations
```{r}
all_pruned_data %>%
  nrow
```

Which are the most significant associations
```{r}
all_pruned_data %>% 
  arrange(P) %>%
  head
```


Which types of endpoints are studied the most frequently?
```{r}
study_details %>%
  select(pubmed_id, phenotype) %>%
  unique %>%
  group_by(phenotype) %>%
  tally %>%
  mutate(percent = n / (study_details %>% select(pubmed_id) %>% unique %>% nrow))
```
How often are individual drugs studied
```{r}
study_details %>%
  mutate(atc_len = nchar(atc)) %>%
  select(pubmed_id, drug, atc, atc_len) %>%
  filter(atc_len == 7) %>%
  group_by(drug) %>%
  tally %>%
  arrange(-n) %>%
  head
```



What is the relationship between how many studies of a drug class there are and the number of significant associations?
```{r message=FALSE, warning=FALSE}
# Number of associations
study_count_data <- all_pruned_data %>%
  group_by(ATC) %>%
  tally %>%
  rename(n_associations = n) %>%
  left_join(atc_totals) %>%
  rename(n_studies = total) %>%
  mutate(ATC_chapter = substr(ATC, 1, 1))

ggplot(study_count_data, aes(n_studies, n_associations)) +
  geom_point(aes(color = ATC_chapter), size = 2.5) +
  theme_Publication(base_size = 12) +
  scale_fill_brewer(palette = "Spectral") +
  ylab("Number of associations") +
  xlab("Number of studies") +
  ggtitle("GWAS yield for drug categories") +
  geom_text_repel(
    data = (study_count_data %>% filter(n_associations > 10)),
    aes(label = ATC),
    box.padding = 0.4
  ) +
  labs(color = "ATC Level 1")

ggsave("figures/atc_study_yield.png",
       width = 6,
       height = 4)
```

### Other miscellaneous analyses

Most studied drugs
```{r}
study_details %>%
  group_by(drug) %>%
  tally %>%
  arrange(-n) %>%
  head
```

Most studied drugs classes
```{r}
study_details %>%
  mutate(ATC = substr(atc, 1, 3)) %>%
  group_by(ATC) %>%
  tally %>%
  arrange(-n) %>%
  head
```

How do PGx GWAS compare to other GWAS in terms of sample size?

```{r}
sample_size_data <- cat_anc %>%
  group_by(STUDY.ACCESSION, PUBMEDID, STAGE, DATE) %>%
  filter(STAGE == "initial") %>%
  summarize(n = sum(NUMBER.OF.INDIVDUALS)) %>%
  mutate(year = year(DATE)) %>%
  mutate(pgx = PUBMEDID %in% study_details$pubmed_id)


ggplot(sample_size_data %>% filter(year < 2021), aes(as.factor(year), n)) +
  geom_boxplot(aes(color = pgx)) +
  scale_y_log10() +
  theme_Publication() +
  ylab("Sample size (log scale)") +
  xlab("Year") +
  scale_colour_Publication() +
  ggtitle("GWAS sample size")
```
What is the median sample size for PGx GWAS vs other GWAS?
```{r}
sample_size_data %>%
  filter(pgx == F) %>%
  filter(!is.na(n)) %>%
    group_by(year, pgx) %>%
  summarize(median(n))

sample_size_data %>%
  ungroup %>%
  filter(year > 2015, year < 2021) %>%
  filter(!is.na(n)) %>%
  group_by(pgx) %>%
  summarize(median(n))
```

Which journals publish the most PGx GWAS?
```{r}
cat_study %>%
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  select(PUBMEDID, JOURNAL) %>%
  unique %>%
  ungroup %>%
  group_by(JOURNAL) %>%
  tally %>%
  arrange(-n) %>%
  head(n=12) %>%
  ggplot(aes(JOURNAL, n)) +
  geom_bar(stat="identity")+
  coord_flip() +
  theme_Publication() +
  ylab("Number of publications") +
  xlab("Journal") +
  ggtitle("Number of publications in top journals")
```



Which genotyping platforms are the most used?
```{r}
cat_full %>%
  select(PUBMEDID, PLATFORM..SNPS.PASSING.QC.) %>%
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  rename(platform = PLATFORM..SNPS.PASSING.QC.) %>%
  unique %>%
  separate(platform, c("platform", NA), sep = "\\[") %>%
  ungroup %>%
  group_by(platform) %>%
  tally %>%
  arrange(-n)
```

Which consortia have led to the most publications?
```{r}
consortia <- read.table("data/consortia_counts.tsv", sep="\t", header=T) 

consortia_ordered <- consortia %>% 
  arrange(-count) %>% 
  filter(consorium != "N/A") %>%
  filter(consorium != "") %>%
  select(consorium) %>%
  pull

consortia %>% 
  arrange(-count) %>% 
  filter(consorium != "N/A") %>%
  filter(consorium != "") %>%
  head(n=12) %>%
  mutate(consorium = ordered(consorium, levels=consortia_ordered)) %>%
  ggplot() +
  geom_bar(aes(consorium, count), stat="identity") +
  theme_Publication() +
  ggtitle("Most used cohorts in PGx GWAS studies") +
  ylab("Number of publications") +
  xlab("Study cohort") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
```

Which studies did the most significant SNPs come from?
```{r}
most_significant_snps <- all_pruned_data %>%
  arrange(P) %>%
  head(n = 20) %>%
  select(rsid) %>%
  pull

cat_full %>%
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  filter(SNPS %in% most_significant_snps) %>%
  select(
    DATE.ADDED.TO.CATALOG,
    PUBMEDID,
    MAPPED_GENE,
    P.VALUE,
    INITIAL.SAMPLE.SIZE,
    DISEASE.TRAIT,
    JOURNAL,
    STUDY,
    SNPS
  ) %>%
  rename(date = DATE.ADDED.TO.CATALOG) %>%
  arrange(MAPPED_GENE, P.VALUE) %>%
  group_by(SNPS) %>%
  filter(P.VALUE == min(P.VALUE))
```



## Publication figures


### Figure 1

Set the palette for this figure
```{r}
two_color_palette = c("#E64B35B2", "#4DBBD5B2")
```


```{r message=FALSE, warning=FALSE}
fig_1a_data <- study_details %>%
  unique %>%
  select(year) %>%
  group_by(year) %>%
  tally %>%
  rename(y = n) %>%
  mutate(facet = "a") %>%
  mutate(facet_name = "# of publications") %>%
  mutate(pgx = "")


#fig_1b_data <- cat_full %>%
fig_1b_data <- cat_study %>%
  select(PUBMEDID, DATE, JOURNAL) %>%
  unique %>%
  mutate(DATE = as.Date(DATE)) %>%
  mutate(year = year(DATE)) %>%
  filter(year < 2021) %>%
  mutate(type = case_when(PUBMEDID %in% study_details$pubmed_id ~ 'PGx',
                          T ~ 'Other')) %>%
  group_by(year, type) %>%
  tally %>%
  pivot_wider(names_from = type, values_from = n) %>%
  mutate(PGx = case_when(is.na(PGx) ~ 0.0, T ~ as.numeric(PGx))) %>%
  mutate(y = (PGx / Other)) %>%
  select(-Other,-PGx) %>%
  mutate(facet = "b") %>%
  mutate(facet_name = "% of all GWAS") %>%
  mutate(pgx = "")

fig_1c_data <- sample_size_data %>%
  ungroup %>%
  filter(year < 2021) %>%
  select(year, n, pgx) %>%
  rename(y = n) %>%
  mutate(y = log10(y)) %>%
  mutate(facet = "c") %>%
  mutate(facet_name = "Sample size") %>%
  mutate(pgx = as.character(pgx))


fig_1_data <- fig_1a_data %>%
  bind_rows(fig_1b_data) %>%
  bind_rows(fig_1c_data) %>%
  filter(year > 2007) %>%
  mutate(year = as.numeric(year))


ggplot(fig_1_data) +
  facet_grid(facet_name ~ ., scales = "free") +
  geom_bar(data = subset(fig_1_data, facet == "a"),
           stat = "identity",
           aes(as.factor(year), y)) +
  geom_point(data = subset(fig_1_data, facet == "b"),
             size = 2,
             aes(as.factor(year), y)) +
  geom_line(data = subset(fig_1_data, facet == "b"), aes(as.factor(year), y)) +
  geom_boxplot(data = subset(fig_1_data, facet == "c"), aes(as.factor(year), y, color =
                                                              pgx)) +
  theme_Publication() +
  ylab("") +
  xlab("Year") +
  ggtitle("Pharmacogenomics GWAS publication statistics") +
  theme(legend.position = "none") +
  scale_x_discrete(breaks = c(2008, 2010, 2012, 2014, 2016, 2018, 2020)) +
  scale_color_manual(values = two_color_palette)

ggsave("figures/figure_1.png", width = 6, height = 6)

```



### Figure 2

Set the palettes
```{r}
eleven_pallete <- c("#332288", "#88CCEE", "#44AA99", "#117733", "#999933", "#DDCC77", "#CC6677", "#882255", "#AA4499", "#DDDDDD")
two_color_palette = c("#BB5566", "#004488")
```

Part A

```{r message=FALSE, warning=FALSE}
top_atc_codes <- all_pruned_data %>%
  group_by(ATC) %>%
  tally %>%
  arrange(-n) %>%
  head(n = 9) %>%
  select(ATC) %>%
  pull

top_atc_codes <- c(top_atc_codes, "Other")


atc_text <- atc_codes %>%
  mutate(ATC = case_when(ATC %in% top_atc_codes ~ ATC, T ~ "Other")) %>%
  mutate(ATC = ordered(ATC, levels = top_atc_codes)) %>%
  select(ATC, atc_description) %>%
  unique %>%
  mutate(
    atc_text = case_when(
      ATC == "B01" ~ "Antithrombotics",
      ATC == "C09" ~ "Renin-angiotensin drugs",
      ATC == "C10" ~ "Lipid modifying",
      ATC == "J05" ~ "Antivirals",
      ATC == "J07" ~ "Vaccines",
      ATC == "L01" ~ "Antineoplastics",
      ATC == "L04" ~ "Immunosuppressants",
      ATC == "N05" ~ "Psycholeptics",
      ATC == "N06" ~ "Psychoanaleptics",
      ATC == "R03" ~ "Obstructive airway",
      ATC == "A07" ~ "Antidiarrheals",
      ATC == "A10" ~ "Diabetes drugs",
      ATC == "C02" ~ "Antihypertensives",
      ATC == "L03" ~ "Immunostimulents",
      ATC == "N03" ~ "Antiepileptics",
      ATC == "C03" ~ "Diuretics",
      T ~ "Other"
    )
  )

atc_text_ordered <- atc_text %>%
  select(atc_text) %>%
  arrange(atc_text) %>%
  filter(atc_text != "Other") %>%
  pull 

atc_text_ordered <- c(atc_text_ordered, "Other")

all_pruned_data %>%
  mutate(ATC = case_when(ATC %in% top_atc_codes ~ ATC, T ~ "Other")) %>%
  mutate(ATC = ordered(ATC, levels = top_atc_codes)) %>%
  left_join(atc_text) %>%
  mutate(atc_text = ordered(atc_text, levels = atc_text_ordered)) %>%
  ggplot(aes(as.factor(year), fill = atc_text)) +
  geom_bar() +
  theme_minimal() +
  ylab("Year") +
  ylab("Count") +
  theme_Publication(base_size = 12) +
  scale_fill_manual(values = eleven_pallete) +
  labs(fill = "ATC") +
  ylab("Number of new associations") +
  xlab("Year") +
  ggtitle("Number of newly discovered PGx associations each year") +
  theme(legend.title = element_blank())

ggsave("figures/figure_2a.png", width = 7, height = 5)
```

Part B

```{r message=FALSE, warning=FALSE}
atc_text <- atc_codes %>%
  filter(ATC %in% all_pruned_data$ATC) %>%
  mutate(ATC = ordered(ATC, levels = atc_order)) %>%
  select(ATC, atc_description) %>%
  unique %>%
  mutate(
    atc_text = case_when(
      ATC == "B01" ~ "Antithrombotics",
      ATC == "C09" ~ "Renin-angiotensin drugs",
      ATC == "C10" ~ "Lipid modifying",
      ATC == "J05" ~ "Antivirals",
      ATC == "J07" ~ "Vaccines",
      ATC == "L01" ~ "Antineoplastics",
      ATC == "L04" ~ "Immunosuppressants",
      ATC == "N05" ~ "Psycholeptics",
      ATC == "N06" ~ "Psychoanaleptics",
      ATC == "R03" ~ "Obstructive airway",
      ATC == "A07" ~ "Antidiarrheals",
      ATC == "A10" ~ "Diabetes drugs",
      ATC == "C02" ~ "Antihypertensives",
      ATC == "L03" ~ "Immunostimulents",
      ATC == "N03" ~ "Antiepileptics",
      ATC == "C03" ~ "Diuretics",
      T ~ "Other"
    )
  )



atc_text_ordered <- atc_text %>%
  
  select(ATC, atc_text) %>%
  unique %>%
  
  mutate(ATC = ordered(ATC, levels = atc_order)) %>%
  
  unique %>%
  filter(atc_text != "Other") %>%
  arrange(ATC) %>%
  select(atc_text) %>%
  pull

atc_text_ordered <- c("Other", atc_text_ordered)

# Number of PGx GWAS publications with significant hits
study_count %>%
  filter(ATC %in% atc_order) %>%
  left_join(atc_text) %>%
  mutate(ATC = ordered(ATC, levels = atc_order)) %>%
  mutate(atc_text = case_when(atc_text %in% atc_text_ordered ~ atc_text, T ~
                                "Other")) %>%
  mutate(atc_text = ordered(atc_text, levels = atc_text_ordered)) %>%
  arrange(-n_studies) %>%
  filter(phenotype %in% c("side effects", "efficacy")) %>%
  mutate(
    phenotype = case_when(
      phenotype == "side effects" ~ "Side effects",
      phenotype == "efficacy" ~ "Efficacy"
    )
  ) %>%
  mutate(n_insig = n_studies - n_studies_sig) %>%
  select(-n_studies) %>%
  pivot_longer(cols = c("n_studies_sig", "n_insig")) %>%
  mutate(plot_name = case_when(
    name == "n_studies_sig" ~ "Significant hits",
    T ~ "No significant hits"
  )) %>%
  ggplot(aes(atc_text, value, fill = plot_name)) +
  geom_bar(stat = "identity") +
  facet_grid(. ~ phenotype) +
  coord_flip() +
  ylab("ATC Code") +
  ylab("Number of studies") +
  ggtitle("Number of PGx GWAS publications") +
  theme_Publication(base_size = 12) +
  scale_fill_Publication() +
  scale_fill_manual(name = "Study findings", values = two_color_palette) +
  labs(fill = "") +
  xlab("ATC Group") +
  ylab("Number of publications") +
  theme(legend.title = element_blank())

ggsave("figures/figure_2b.png", width = 6, height = 6)
```

### Figure 3

The complete code for Figure 3 needs to be rewritten. Please open an issue if you would like it and it's not here yet.
```{r}
cat_anc_modified <- read.table("data/cat_ancestry_modified.tsv", sep=",", header=T)

```

```{r}
anc_histrogram_data <- read.table("data/broad_ancestral_bystudytype.csv", sep=",", header=T)
anc_histrogram_data %>% select(Broader, Discovery, Replication) %>%
  pivot_longer(c(Discovery, Replication)) %>%
  ggplot(aes(Broader, value/100)) +
  geom_histogram(stat="identity", aes(fill=Broader)) +
  facet_grid(name~.) +
  theme_Publication() +
  ggtitle("Ancestral makeup of PGx GWAS studies") +
  ylab("Percent of total") +
  scale_y_continuous(labels=percent) +
  xlab("Ancestral population") +
  theme(axis.text.x = element_blank())
  
  
```

### Figure 4


Set the palette

```{r}
palette_25 <- c(
  "dodgerblue2", "#E31A1C", # red
  "green4",
  "#6A3D9A", # purple
  "#FF7F00", # orange
  "black", "gold1",
  "skyblue2", "#FB9A99", # lt pink
  "palegreen2",
  "#CAB2D6", # lt purple
  "#FDBF6F", # lt orange
  "gray70", "khaki2",
  "maroon", "orchid1", "deeppink1", "blue1", "steelblue4",
  "darkturquoise", "green1", "yellow4", "yellow3",
  "darkorange4", "brown"
)
```


The figure in the publication was manually cropped and rotated
```{r message=FALSE, warning=FALSE}
consortia_ordered <- consortia %>%
  unique %>%
  group_by(consortium) %>%
  tally %>%
  rename(count = n) %>%
  arrange(-count) %>%
  filter(consortium != "N/A") %>%
  select(consortium) %>%
  pull


f4a_data <- consortia %>%
  filter(consortium %in% consortia_ordered[0:17]) %>%
  unique %>%
  mutate(consortium = ordered(consortium, levels = consortia_ordered)) %>%
  group_by(consortium, atc_text) %>%
  tally %>%
  unique %>%
  mutate(year = 2000) %>%
  mutate(facet = "# of GWAS") %>%
  filter(!is.na(atc_text))


f4b_data <- consortia %>%
  filter(consortium %in% consortia_ordered[0:17]) %>%
  unique %>%
  mutate(consortium = ordered(consortium, levels = consortia_ordered)) %>%
  left_join(
    cat_full %>% select(PUBMEDID, DATE) %>% unique %>% rename(pmid = PUBMEDID) %>% mutate(pmid =
                                                                                            as.character(pmid))
  ) %>%
  mutate(year = as.numeric(format(DATE, '%Y'))) %>%
  select(year, pmid, consortium) %>%
  unique %>%
  group_by(year, consortium) %>%
  tally %>%
  mutate(atc_text = "") %>%
  mutate(facet = "Publications over time")



f4_data <- bind_rows(f4a_data, f4b_data)


ggplot(f4_data) +
  facet_grid(facet ~ ., scales = "free", switch = "both") +
  geom_bar(
    data = subset(f4_data, facet == "# of GWAS"),
    aes(consortium, n, fill = atc_text),
    stat = "identity"
  ) +
  geom_point(data = subset(f4_data, facet == "Publications over time"),
             aes(consortium, year, size = n)) +
  theme_Publication() +
  theme(
    axis.text.x = element_text(
      angle = 90,
      vjust = 0.5,
      hjust = 1
    ),
    legend.title = element_blank(),
    strip.placement = "outside",
    axis.text.y = element_text(angle = 90),
    axis.title.x = element_text(angle = 180)
  ) +
  scale_y_continuous(position = "right") +
  scale_fill_manual(values = palette_25) +
  xlab("Data source") +
  ylab("")

ggsave("figure_4.png", width = 8, height = 8)

```





### Figure 5

Set the palette

```{r}
four_color_palette <- rev(c("#4477AA", "#228833", "#CCBB44", "#EE6677"))
```



```{r message=FALSE, warning=FALSE}
# Laod the data
vip_gene_coords <-
  read.table("data/pgx_vip_hg38.extra.bed",
             sep = "\t",
             header = T)
names(vip_gene_coords) <- c("chr", "start", "end", "gene", "tier")

# Extract the GWAS hits within each gene
vip_gene_hits <- data.frame()
for (i in 1:nrow(vip_gene_coords)) {
  chr = vip_gene_coords[i, ]$chr
  start = vip_gene_coords[i, ]$start - 50000
  end = vip_gene_coords[i, ]$end + 50000
  gene = vip_gene_coords[i, ]$gene
  
  hits <- pgx_gwas_hits %>%
    filter(CHR == chr) %>%
    filter(BP >= start, BP <= end) %>%
    group_by(atc) %>%
    filter(P == min(P)) %>%
    select(P, atc, phenotype, PUBMEDID) %>%
    mutate(gene_name = gene)
  
  vip_gene_hits <- bind_rows(vip_gene_hits, hits)
}

# Set the order for the ATC codes in the plot
atc_order <- vip_gene_hits %>%
  mutate(len = str_length(atc)) %>%
  #filter(len == 7) %>%
  rename(ATC = atc) %>%
  left_join(atc_codes) %>%
  select(ATC, atc_description) %>%
  unique %>%
  arrange(ATC) %>%
  mutate(
    atc_description = case_when(
      atc_description == "beta-Adrenergic Blocking Agents" ~ "beta blockers",
      atc_description == "Selective serotonin reuptake inhibitors" ~ "SSRIs",
      atc_description == "Sulfur-containing imidazole derivatives, antithyroid preparations" ~ "imidazoles",
      atc_description == "Thiouracil antithyroid preparations" ~ "thiouracils",
      atc_description == "Bisphosphonate drugs affecting bone structure and mineralization" ~ "bisphosphonates",
      atc_description == "Sulfonylureas for blood glucose lowering" ~ "sulfonylureas",
      atc_description == "Carboxamide derivatives, antiepileptics" ~ "carboxamides",
      atc_description == "hepatitis B, purified antigen" ~ "hepatitis B vaccine",
      atc_description == "smallpox, live attenuated" ~ "smallpox vaccine",
      ATC == "N03AF" ~ "carboxamide derivatives",
      T ~ atc_description
    )
  ) %>%
  select(atc_description) %>%
  pull

# This big ugly snippet formats the data and generates the figure. Apologies...
vip_gene_hits %>%
  mutate(len = str_length(atc)) %>%
  #filter(len == 7) %>%
  rename(ATC = atc) %>%
  filter(!ATC %in% c(
    "N05A",
    "L04AA",
    "C10A",
    "C10AA",
    "N/A",
    "L01",
    "J07BC",
    "J05A",
    "C03AA"
  )) %>%
  filter(PUBMEDID != 23459443) %>%
  left_join(atc_codes) %>%
  mutate(
    atc_description = case_when(
      atc_description == "beta-Adrenergic Blocking Agents" ~ "beta blockers",
      atc_description == "Selective serotonin reuptake inhibitors" ~ "SSRIs",
      atc_description == "Sulfur-containing imidazole derivatives, antithyroid preparations" ~ "imidazoles",
      atc_description == "Thiouracil antithyroid preparations" ~ "thiouracils",
      atc_description == "Bisphosphonate drugs affecting bone structure and mineralization" ~ "bisphosphonates",
      atc_description == "Sulfonylureas for blood glucose lowering" ~ "sulfonylureas",
      atc_description == "Carboxamide derivatives, antiepileptics" ~ "carboxamides",
      atc_description == "hepatitis B, purified antigen" ~ "hepatitis B vaccine",
      atc_description == "smallpox, live attenuated" ~ "smallpox vaccine",
      ATC == "N03AF" ~ "carboxamide derivatives",
      T ~ atc_description
    )
  ) %>%
  mutate(atc_description = ordered(atc_description, levels = rev(atc_order))) %>%
  mutate(
    phenotype = case_when(
      phenotype == "biomarker" ~ "Biomarker",
      phenotype == "efficacy" ~ "Response",
      phenotype == "metabolism" ~ "Metabolism",
      phenotype == "other" ~ "Response",
      phenotype == "side effects" ~ "ADRs"
    )
  ) %>%
  left_join(vip_gene_coords %>% select(gene, tier) %>% rename(gene_name =
                                                                gene)) %>%
  mutate(
    tier = case_when(
      tier == "Tier 1" ~ "T1",
      tier == "Tier 2" ~ "T2",
      tier == "Cancer Genome" ~ "CG",
      tier == "Other" ~ "HLA"
    )
  ) %>%
  mutate(tier = ordered(tier, levels = c("T1", "T2", "CG", "HLA"))) %>%
  mutate(significant = case_when(P < 5e-8 ~ TRUE,
                                 T ~ F)) %>%
  mutate(ATC_chapter = substr(ATC, 1, 1)) %>%
  left_join(atc_codes %>% rename(ATC_chapter = ATC)) %>%
  
  ggplot() +
  geom_point(aes(
    gene_name,
    ATC,
    color = phenotype,
    size = -log10(P),
    alpha = significant
  )) +
  facet_grid(ATC_chapter ~ tier, scales = "free", space = "free") +
  theme_Publication() +
  theme(axis.text.x = element_text(
    angle = 90,
    vjust = 0.5,
    hjust = 1,
    face = "italic"
  )) +
  scale_color_manual(values = four_color_palette) +
  xlab("Gene") +
  ylab("Drug") +
  ggtitle("VIP PGx gene-drug associations in GWAS") +
  scale_alpha_discrete(range = c(0.5, 1))


ggsave("figures/figure_5.png", width = 8.5, height = 8)
```






### Supplemental figures


Counts for ADR vs Efficacy studies
```{r}
study_sample_sizes <- study_details %>%
  left_join(sample_size_data %>% select(PUBMEDID, n) %>% rename(pubmed_id = PUBMEDID)) %>%
  filter(phenotype %in% c("efficacy", "side effects")) %>%
  filter(!is.na(n)) %>%
  mutate(phenotype = case_when(phenotype == "efficacy" ~ "Response",
                               phenotype == "side effects" ~ "ADRs"))

ggplot(study_sample_sizes) +
  geom_boxplot(aes(phenotype, n)) +
  theme_Publication() +
  scale_y_continuous(trans='log10') +
  ylab("Sample size (log scale)") +
  xlab("Studied phenotype") +
  ggtitle("Sample size of PGx GWAS for Efficacy or ADRs")
```

```{r}
ggplot(study_sample_sizes %>% filter(!is.na(year))) +
  geom_boxplot(aes(as.factor(year), n, fill=phenotype)) +
  theme_Publication() +
  scale_y_continuous(trans='log10') +
  ylab("Sample size (log scale)") +
  xlab("Studied phenotype") +
  ggtitle("Sample size of PGx GWAS for Efficacy or ADRs") +
  ggsave("figures/efficacy_vs_adr_sample_size.png", width=8, height=4)
```







```{r message=FALSE, warning=FALSE}
cat_full %>% 
  #filter(PUBMEDID == 32131869) %>% 
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  mutate(significant = case_when(P.VALUE <= 5e-08 ~ 1, T~0)) %>%
  #mutate(significant = ordered(significant, levels=c(F, T))) %>%
  select(PUBMEDID, significant) %>%
  group_by(PUBMEDID) %>%
  select(PUBMEDID, significant) %>%
  left_join(study_sample_counts) %>%
  #unique %>%
  left_join(study_details %>% rename(PUBMEDID=pubmed_id) %>% select(PUBMEDID, phenotype)) %>%
  group_by(STUDY.ACCESSION, total_subjects, phenotype) %>%
  summarize(n_hits = sum(significant)) %>%
  unique %>%
  ggplot() +
  geom_point(aes(total_subjects, n_hits, color=phenotype), alpha=0.5) +
  ylim(0,50) +
  theme_Publication() +
  ggtitle("PGx GWAS sample size compared to\nnumber of significant associations") +
  ylab("# significant associations") +
  xlab("# subjects (log scale)") +
  scale_color_brewer(palette = "Spectral") +
  scale_x_log10()

ggsave("figures/sample_size_vs_hits.png", width=8, height=4)
```

```{r message=FALSE, warning=FALSE}
hit_counts <- cat_full %>% 
  #filter(PUBMEDID == 32131869) %>% 
  filter(PUBMEDID %in% study_details$pubmed_id) %>%
  mutate(significant = case_when(P.VALUE <= 5e-08 ~ 1, T~0)) %>%
  #mutate(significant = ordered(significant, levels=c(F, T))) %>%
  select(PUBMEDID, significant) %>%
  group_by(PUBMEDID) %>%
  select(PUBMEDID, significant) %>%
  left_join(study_sample_counts) %>%
  #unique %>%
  left_join(study_details %>% rename(PUBMEDID=pubmed_id) %>% select(PUBMEDID, phenotype, year)) %>%
  group_by(PUBMEDID, STUDY.ACCESSION, total_subjects, phenotype, year) %>%
  summarize(n_hits = sum(significant)) %>%
  arrange(-n_hits) %>%
  ungroup %>%
  #select(PUBMEDID, n_hits) %>%
  unique 

cor(log10(hit_counts$total_subjects), hit_counts$n_hits)
```


